### PR TITLE
🌱 Generate events for topology changes

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR improves the topology controller so events will be generated for each change applied to the managed objects

### CODE REVIEW

File: controllers/topology/internal/mergepatch/mergepatch.go [hunk 1]
```diff
@@ -293,5 +293,5 @@ func (h *Helper) Patch(ctx context.Context) error {
 	log := ctrl.LoggerFrom(ctx)
 	log.V(5).Info("Patching object", "Patch", string(h.patch))
 
-	return h.client.Patch(ctx, h.original, client.RawPatch(types.MergePatchType, h.patch))
+	return h.client.Patch(ctx, h.original.DeepCopyObject().(client.Object), client.RawPatch(types.MergePatchType, h.patch))
```
Comment (fabriziopandini) [inline controllers/topology/internal/mergepatch/mergepatch.go:296]: This is for preventing side effects on the object

File: controllers/topology/reconcile_state_test.go [hunk 2]
```diff
@@ -678,8 +682,8 @@ func TestReconcileMachineDeployments(t *testing.T) {
 		{
 			name:    "Should update MachineDeployment with InfrastructureMachineTemplate and BootstrapTemplate without rotation",
 			current: []*scope.MachineDeploymentState{md4m},
-			desired: []*scope.MachineDeploymentState{md4mWithRotatedTemplates},
-			want:    []*scope.MachineDeploymentState{md4m},
+			desired: []*scope.MachineDeploymentState{md4mWithInPlaceUpdatedTemplates},
+			want:    []*scope.MachineDeploymentState{md4mWithInPlaceUpdatedTemplates},
```
Comment (fabriziopandini) [inline controllers/topology/reconcile_state_test.go:686]: This test was wrong, but the side effect in mergepatch.Patch was hiding the error

File: controllers/topology/cluster_controller.go [hunk 3]
```diff
@@ -108,6 +110,7 @@ func (r *ClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_
 		// Error reading the object - requeue the request.
 		return ctrl.Result{}, err
 	}
+	cluster.Kind = "Cluster"
```
Comment (vincepri) [inline controllers/topology/cluster_controller.go:113]: Why is this required?

File: controllers/topology/internal/mergepatch/mergepatch.go (hunk 1 above)
Comment (vincepri) [reply controllers/topology/internal/mergepatch/mergepatch.go:296]: Side effect where?

File: controllers/topology/reconcile_state.go [hunk 4]
```diff
@@ -288,44 +325,77 @@ func calculateMachineDeploymentDiff(current, desired map[string]*scope.MachineDe
...
-	if current == nil {
-		log.Infof("Creating %s", tlog.KObj{Obj: desired})
-		if err := r.Client.Create(ctx, desired.DeepCopy()); err != nil {
-			return errors.Wrapf(err, "failed to create %s", tlog.KObj{Obj: desired})
+	if in.current == nil {
+		log.Infof("Creating %s", tlog.KObj{Obj: in.desired})
+		if err := r.Client.Create(ctx, in.desired.DeepCopy()); err != nil {
+			return errors.Wrapf(err, "failed to create %s", tlog.KObj{Obj: in.desired})
 		}
+		r.recorder.Eventf(in.cluster, corev1.EventTypeNormal, "TopologyCreate", "Created %q", tlog.KObj{Obj: in.desired})
```
Comment (enxebre) [inline controllers/topology/reconcile_state.go:350]: does it worth having TopologyCreate/TopologyDelete/TopologyUpdate into constants?

File: controllers/topology/cluster_controller.go (hunk 3 above)
Comment (enxebre) [inline controllers/topology/cluster_controller.go:113]: curious why is this needed?

Comment (sbueringer) [reply controllers/topology/cluster_controller.go:113]: I think the idea is to set the Kind so the event recorder can set it correctly when creating the event.

But if I saw correctly the event recorder will try to get the kind from the scheme as a fallback:
https://github.com/kubernetes/client-go/blob/master/tools/reference/ref.go#L59-L72

But I might have looked at the wrong code as I just was clicking through endless interfaces, and didn't try it with the debugger or something.

File: controllers/topology/internal/mergepatch/mergepatch.go (hunk 1 above)
Comment (fabriziopandini) [reply controllers/topology/internal/mergepatch/mergepatch.go:296]: h.original gets modified by client.Patch

File: controllers/topology/cluster_controller.go (hunk 3 above)
Comment (fabriziopandini) [reply controllers/topology/cluster_controller.go:113]: The problem is our KObj object reporting a empty kind for the Cluster (it does not have fall back to schema)

File: controllers/topology/internal/mergepatch/mergepatch.go (hunk 1 above)
Comment (sbueringer) [reply controllers/topology/internal/mergepatch/mergepatch.go:296]: Should we add a comment?

File: controllers/topology/reconcile_state.go [hunk 5]
```diff
@@ -136,6 +153,7 @@ func (r *ClusterReconciler) reconcileCluster(ctx context.Context, s *scope.Scope
 	if err := patchHelper.Patch(ctx); err != nil {
 		return errors.Wrapf(err, "failed to patch %s", tlog.KObj{Obj: s.Current.Cluster})
 	}
+	r.recorder.Eventf(s.Current.Cluster, corev1.EventTypeNormal, "TopologyUpdate", "Updated %q", tlog.KObj{Obj: s.Current.Cluster})
```
Comment (sbueringer) [inline controllers/topology/reconcile_state.go:156]: I'm not sure about the conventions around event reasons, but it looks like in other cases we often have `Failed<>` and a `Successful<>` event.

This could mean:
* let's prefix the reason with Successful or Failed
* let's also publish an event in the error case

Comment (sbueringer) [reply controllers/topology/reconcile_state.go:156]: Also not sure if it's better to use something like `SuccessfulTopologyUpdate` or `SuccessfulClusterUpdate`
(I'm fine with both, would be interesting if there are some Kubernetes upstream conventions)

**Review by sbueringer — COMMENTED** [review]
Looks great!

File: controllers/topology/internal/mergepatch/mergepatch.go (hunk 1 above)
Comment (fabriziopandini) [reply controllers/topology/internal/mergepatch/mergepatch.go:296]: Done!

File: controllers/topology/reconcile_state.go (hunk 4 above)
Comment (fabriziopandini) [reply controllers/topology/reconcile_state.go:350]: Done!

File: controllers/topology/cluster_controller.go (hunk 3 above)
Comment (fabriziopandini) [reply controllers/topology/cluster_controller.go:113]: see https://github.com/kubernetes-sigs/cluster-api/pull/5631#discussion_r746945426

File: controllers/topology/reconcile_state.go (hunk 5 above)
Comment (fabriziopandini) [reply controllers/topology/reconcile_state.go:156]: > Also not sure if it's better to use something like SuccessfulTopologyUpdate or SuccessfulClusterUpdate
(I'm fine with both, would be interesting if there are some Kubernetes upstream conventions)

I'm personally -1 with this one, given that there is already event type normal/warning
Also if I look at Kubernetes events after creating a cluster, Successful/Failed prefix is not usual and the only place where this is usual are MachineDeployment/MachineSet :-(

```
6m3s        Normal   Starting                  node/capi-test-control-plane                  Starting kubelet.
6m3s        Normal   NodeHasSufficientMemory   node/capi-test-control-plane                  Node capi-test-control-plane status is now: NodeHasSufficientMemory
6m3s        Normal   NodeHasNoDiskPressure     node/capi-test-control-plane                  Node capi-test-control-plane status is now: NodeHasNoDiskPressure
6m3s        Normal   NodeHasSufficientPID      node/capi-test-control-plane                  Node capi-test-control-plane status is now: NodeHasSufficientPID
6m3s        Normal   NodeAllocatableEnforced   node/capi-test-control-plane                  Updated Node Allocatable limit across pods
5m51s       Normal   Starting                  node/capi-test-control-plane                  Starting kubelet.
5m51s       Normal   NodeHasSufficientMemory   node/capi-test-control-plane                  Node capi-test-control-plane status is now: NodeHasSufficientMemory
5m51s       Normal   NodeHasNoDiskPressure     node/capi-test-control-plane                  Node capi-test-control-plane status is now: NodeHasNoDiskPressure
5m51s       Normal   NodeHasSufficientPID      node/capi-test-control-plane                  Node capi-test-control-plane status is now: NodeHasSufficientPID
5m51s       Normal   NodeAllocatableEnforced   node/capi-test-control-plane                  Updated Node Allocatable limit across pods
5m39s       Normal   RegisteredNode            node/capi-test-control-plane                  Node capi-test-control-plane event: Registered Node capi-test-control-plane in Controller
5m34s       Normal   Starting                  node/capi-test-control-plane                  Starting kube-proxy.
5m21s       Normal   NodeReady                 node/capi-test-control-plane                  Node capi-test-control-plane status is now: NodeReady
5s          Normal   SuccessfulCreate          machineset/my-cluster1-md1-rdtlv-864ffd466c   Created machine "my-cluster1-md1-rdtlv-864ffd466c-hb74t"
5s          Normal   SuccessfulCreate          machinedeployment/my-cluster1-md1-rdtlv       Created MachineSet "my-cluster1-md1-rdtlv-864ffd466c"
5s          Normal   TopologyCreate            cluster/my-cluster1                           Created "DockerCluster/my-cluster1-fmjqk"
5s          Normal   TopologyCreate            cluster/my-cluster1                           Created "DockerMachineTemplate/my-cluster1-control-plane-shgx4"
5s          Normal   TopologyCreate            cluster/my-cluster1                           Created "KubeadmControlPlane/my-cluster1-nzmzw"
5s          Normal   TopologyUpdate            cluster/my-cluster1                           Updated "Cluster/my-cluster1"
5s          Normal   TopologyCreate            cluster/my-cluster1                           Created "DockerMachineTemplate/my-cluster1-md1-infra-x5dgv"
5s          Normal   TopologyCreate            cluster/my-cluster1                           Created "KubeadmConfigTemplate/my-cluster1-md1-bootstrap-x29kb"
5s          Normal   TopologyCreate            cluster/my-cluster1                           Created "MachineDeployment/my-cluster1-md1-rdtlv"
```

> let's also publish an event in the error case

I'm +1 but I would defer to a follow up PR given that I would like to have consistency with conditions

**Comment by sbueringer** [comment]
@fabriziopandini @vincepri Where are we on this PR? Would be nice to get it in to be able to validate it as part of some manual testing.

**Comment by fabriziopandini** [comment]
From my point of view we are ready to go.
It provides a minimal set of event reporting changes to topology-owned objects. We can then eventually iterate in the future e.g. adding events for topology reconcile errors, more specific events for version upgrades, rebase etc.

File: controllers/topology/cluster_controller.go (hunk 3 above)
Comment (vincepri) [reply controllers/topology/cluster_controller.go:120]: Could we actually use the scheme here to set the GVK?

**Review by vincepri — APPROVED** [review]
/approve
/lgtm
